vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc

```diff
@@ -69,7 +90,117 @@ void AdConversions::ProcessQueue() {
   StartTimer(ad_conversion);
 }
 
-void AdConversions::AddToQueue(
+bool AdConversions::OnTimer(
+    const uint32_t timer_id) {
+  if (timer_id != timer_id_) {
+    return false;
+  }
+
+  timer_id_ = 0;
+
+  DCHECK(!queue_.empty());
+  if (queue_.empty()) {
+    return true;
+  }
+
+  AdConversionQueueItemInfo ad_conversion = queue_.front();
+  ProcessQueueItem(ad_conversion);
+
+  return true;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void AdConversions::OnGetAdConversions(
+    const std::string& url,
+    const Result result,
+    const AdConversionList& ad_conversions) {
+  if (result != SUCCESS) {
+    BLOG(ERROR) << "Failed to check ad conversions";
+    return;
+  }
+
+  std::deque<AdHistory> ads_history = client_->GetAdsShownHistory();
+  ads_history = FilterAdsHistory(ads_history);
+  ads_history = SortAdsHistory(ads_history);
+
+  AdConversionList new_ad_conversions = ad_conversions;
+  new_ad_conversions = FilterAdConversions(url, new_ad_conversions);
+  new_ad_conversions = SortAdConversions(new_ad_conversions);
+
+  for (const auto& ad_conversion : new_ad_conversions) {
+    for (const auto& ad : ads_history) {
+      auto ad_conversion_history = client_->GetAdConversionHistory();
+      if (ad_conversion_history.find(ad_conversion.creative_set_id) !=
+          ad_conversion_history.end()) {
+        // Creative set id has already been converted
+        continue;
+      }
+
+      if (ad_conversion.creative_set_id != ad.ad_content.creative_set_id) {
+        // Creative set id does not match
+        continue;
+      }
+
+      const base::Time observation_window = base::Time::Now() -
+          base::TimeDelta::FromDays(ad_conversion.observation_window);
+      const base::Time time = Time::FromDoubleT(ad.timestamp_in_seconds);
+      if (observation_window > time) {
+        // Observation window has expired
+        continue;
+      }
+
+      BLOG(INFO) << "Ad conversion for " << url << " with "
+          << ad_conversion.creative_set_id << " creative set id for "
+              << std::string(ad_conversion.type);
+
+      AddItemToQueue(ad.ad_content.creative_instance_id,
```

Comment:
is this synchronous? I.e. if we have an ad with the same creative_set_id as the ad that just converted will it return in L137? Is this also the reason why we can't pull L133 outside the loop?

Reply:
Yes to both, thanks